api/src/main/java/org/apache/iceberg/io/FileIOMetrics.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * Optional interface for FileIO implementations to report metrics related
+ * to read/write operations.
+ */
+public interface FileIOMetrics extends Serializable {
+  String READ_BYTES = "read.bytes";
+  String READ_OPERATIONS = "read.operations";
+  String WRITE_BYTES = "write.bytes";
+  String WRITE_OPERATIONS = "write.operations";
+
+  default void initialize(Map<String,String> properties) {}
+
+  interface Counter<T> {
+    void increment(T amount);
+  }
+
+  /**
+   * Get a named counter of a specific type.  Metric implementations may impose
+   * restrictions on what types are supported for specific counters.
+   *
+   * @param name name of the metric
+   * @param type numeric type of the counter value
+   * @return a counter implementation
+   */
+  <T> Counter<T> counter(String name, Class<T> type);
```

Comment:
For now, I'm only providing a Counter interface due to the first implementation being Hadoop metrics, which don't require other meters (e.g. Timer, Gauge).  However, this could be expanded upon similarly if those use cases emerge.

Reply:
I guess for a super simple representation of metrics this might be enough. Once requirements get a little bit more sophisticated (by wanting to have Tag support for metrics), then it's possible that maintaining an own metric implementation gets more difficult.

Reply:
I think the implementation could still add tags based on the metric name and property based initialization.  I don't want to overcomplicate or be too opinionated about the implementation.  To your point, we could always expand upon the implementation, but the first use case (hadoop fs reporting) is really limited in scope.

Reply:
I agree on both accounts.

Tags can be really useful and the biggest complaint by Flink users when I administered Flink clusters was not being able to used tagged metrics (though not all metrics systems support them directly - many still use the one metric maps to one key value pair basically). However, we eventually were able to use tagged metrics directly via the prometheus Java library directly within Flink by some configuration that Flink's metric system provided.

I imagine Micrometer might have also been able to do the same (if it was exporting metrics in a tagged format).

I agree that for the first use case, just a simple Counter is sufficient and given how limited the hadoop fs metrics reporting is in scope (and how requested those metrics are from S3FileIO), it would be better to not complicate things initially.